proposals/0000-linear-types.rst

```diff
@@ -0,0 +1,2245 @@
+the same multiplicity as the intermediate variable).
+
+::
+
+  f mxy = case_1 mxy of
+    Just xy -> case_1 xy of
+      (x, y) -> …
+
+Typechecking equations
+~~~~~~~~~~~~~~~~~~~~~~
+
+In a definition with multiple equations, each equation is typechecked
+independently.
+
+Let us see an equation as a list of (typed) binders (*i.e.* patterns)
+and a right-hand side. Each binder has a multiplicity, which is
+provided by the signature. If there is no signature, the initial
+multiplicity of each binder is ω instead.
+
+Let us consider a judgement ``Γ ⊢ (b1 :(π1) A1) … (bn :(πn) An) → u : B``
+
+- ``Γ ⊢ u : B ⟹ Γ ⊢ → u : B``
+- ``Γ, x :(π) A ⊢ (b1 :(π1) A1) … (bn :(πn) An) → u : B ⟹ Γ ⊢ (x
+  :(π) A) (b1 :(π1) A1) … (bn :(πn) An) → u : B``
+- ``Γ ⊢ (p1 :(πρ1) C1) … (pn :(πρn) Cn) (b1 :(π1) A1) … → u : B ⟹ Γ ⊢ (c p1 …
+  pn :(π) D) (b1 :(π1) A1) … → u : B``, for ``c : C1 :ρ1-> … Cn :ρn->
+  D``, a constructor (notice how ``π`` flows down into the fields of ``c``)
+- ``Γ ⊢ (b1 :(π1) A1) … → u : B ⟹ Γ ⊢ (_ :(π) C) (b1 :(π1) A1) … → u :
+  B``, if ``π=ω``
+
+
+Unboxed data types
+------------------
+
+GHC supports unboxed data types such as ``(#,#)`` (unboxed pair) and
+``(#|#)`` (binary unboxed sum). The proposal treats them as their boxed
+equivalent (``(,)`` and ``Either``, respectively, for these two
+examples): the constructors are linear (and case can have various
+multiplicities).
+
+There is no current proposed syntax for unboxed data types of mixed
+multiplicity, though the `Unlifted data types proposal
+<https://ghc.haskell.org/trac/ghc/wiki/UnliftedDataTypes>`_ (if
+extended to unboxed data types as well), could provide a
+solution. Mixed-multiplicity unboxed records are, however, required
+internally (see `The Core corner`_): they simply don't have a syntax
+yet.
+
+Records and projections
+-----------------------
+
+.. _Records:
+
+Records constructors, with Haskell98 syntax, are linear. That is, in
+
+::
+
+   data R = R {f1 :: A1, … fn :: An}
+
+we have ``R :: A1 ->. … ->. An ->. R``.
+
+Mixed-multiplicity records can be defined using the syntax for
+annotating binders with multiplicity
+
+::
+
+  data R' = R' { f1 ::('Omega) A1, f2 ::('One) A2, f3 :: A3 }
+
+Then ``R' :: A1 -> A2 ->. A3 ->. R`` (that is, fields with no explicit
+annotation are linear).
+
+Record patterns act like tuple patterns, but some fields can be
+omitted. A field can be omitted only if the resolved multiplicity for
+this field is ω.
+
+::
+
+  foo :: R' ->. A
+  foo {f2=x, f3=y} = … -- permitted as f1 has multiplicity ω
+  foo {f2=x} = … -- rejected as f3 is omitted and has multiplicity 1
+
+  foo :: R' -> A -- non-linear function!
+  foo {f2=x} = … -- permitted because the context has multiplicity ω,
+                 -- hence the resolved multiplicity of f3 is ω.
+
+Projections take an *unrestricted* record as argument: ``f1 :: R ->
+A1`` (because otherwise the other fields would not be consumed). There
+is an exception to this rule: if all the other fields are
```

Comment:
22. I don't think it's necessarily to generalize the record-selector exception. I was just testing my understanding of it all. But I do think the proposal should mention that the exception applies only to single-argument types.